### PR TITLE
docs: add searching within sidebar

### PR DESCRIPTION
**Pull request by onerandomusername**
## Summary

Adds some searching of the objects within the sidebar. This should make it easier to find a specific object when you already know a portion of the name.

Thanks to @ItsAleph for the idea and @bast0006 for help with the implementation.

(still has some css stuff to look at but made the pr for the rtd preview)

## Checklist

- [ ] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [ ] I have formatted the code properly by running `pdm lint`
    - [ ] I have type-checked the code by running `pdm pyright`
- [ ] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [x] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: docs/_static/sidebar.js [hunk 1]
```diff
@@ -121,3 +121,29 @@ document.addEventListener('DOMContentLoaded', () => {
...
+      while (el) {
+        el.style.display = "";
+
+        el = el.parentNode.closest("#sidebar li");
+      }
+    } else {
+      li[i].style.display = "none";
+    }
+  }
+}
```
Comment (shiftinv) [inline docs/_static/sidebar.js:149]:
```suggestion
function sidebarSearch() {
  const filter = document.getElementById('sidebar-search-box').value.toUpperCase();
  const items = document.querySelectorAll('#sidebar li')

  // Loop through all list items, and hide those who don't match the search query
  for (const item of items) {
    const a = item.querySelector("a");
    const itemText = a.textContent || a.innerText;
    if (itemText.toUpperCase().indexOf(filter) > -1) {
      // we need to convert all of the parents back into visible mode
      let el = item;
      while (el) {
        el.style.display = "";
        el = el.parentNode.closest("#sidebar li");
      }
    } else {
      item.style.display = "none";
    }
  }
}
```
smol refactor

File: docs/_static/sidebar.js [hunk 2]
```diff
@@ -121,3 +121,29 @@ document.addEventListener('DOMContentLoaded', () => {
...
+  var input, filter, ul, li, a, i, txtValue;
+  input = document.getElementById('sidebar-search-box');
+  filter = input.value.toUpperCase();
+  ul = document.getElementById("sidebar");
+  li = ul.querySelectorAll('li');
+
+  // Loop through all list items, and hide those who don't match the search query
+  for (i = 0; i < li.length; i++) {
+    a = li[i].getElementsByTagName("a")[0];
+    txtValue = a.textContent || a.innerText;
```
Comment (shiftinv) [inline docs/_static/sidebar.js:136]: Why fall back to `innerText` here?

File: changelog/963.doc.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+
+Add a sidebar searchbox for searching within the sidebar on the API Reference pages.
```
Comment (shiftinv) [inline changelog/963.doc.rst:2]:
```suggestion
Add a sidebar searchbox for searching within the sidebar on the API Reference pages.
```

File: docs/_static/style.css [hunk 4]
```diff
@@ -567,6 +567,10 @@ header form.search {
   #hamburger-toggle {
     left: 8px;
   }
+
+  .sidebar-search-wrapper {
+    width: fit-content;
```
Comment (shiftinv) [inline docs/_static/style.css:572]: this always gets overridden by `width: 200px` below

File: docs/_static/style.css [hunk 5]
```diff
@@ -669,6 +673,33 @@ aside .material-icons {
   align-items: baseline;
 }
 
+/* sidebar search box stuff */
+
+.sidebar-search-wrapper {
+  display: flex;
+  align-items: stretch;
```
Comment (shiftinv) [inline docs/_static/style.css:680]: flex layouts won't apply with a fixed `width:` set

File: docs/_templates/layout.html [hunk 6]
```diff
@@ -122,6 +122,11 @@
     {%- endfor %}
   </div>
   <div id="sidebar">
+    {% if pagename == 'api' or '/api' in pagename %}
+    <div class="sidebar-search-wrapper">
+      <input type="text" id="sidebar-search-box" onkeyup="sidebarSearch()" placeholder="Search API Reference.." autocomplete=off>
```
Comment (shiftinv) [inline docs/_templates/layout.html:127]:
```suggestion
      <input type="text" id="sidebar-search-box" onkeyup="sidebarSearch()" placeholder="Filter..." autocomplete=off>
```

File: docs/_templates/layout.html [hunk 7]
```diff
@@ -122,6 +122,11 @@
     {%- endfor %}
   </div>
   <div id="sidebar">
+    {% if pagename == 'api' or '/api' in pagename %}
```
Comment (shiftinv) [inline docs/_templates/layout.html:125]:
```suggestion
    {% if pagename.split("/")[-1] == "api" %}
```

Comment (onerandomusername) [reply docs/_templates/layout.html:125]: How this currently is makes it future proofed against #722.

Comment (onerandomusername) [reply docs/_templates/layout.html]: optimised in https://github.com/DisnakeDev/disnake/pull/963/commits/9dcffe34b474a846ceb0bd6fb218ca8c42b780b2 and kept it future proofed

File: docs/_static/sidebar.js (hunk 2 above)
Comment (onerandomusername) [reply docs/_static/sidebar.js]: resolved by refactoring

File: docs/_static/style.css (hunk 5 above)
Comment (onerandomusername) [reply docs/_static/style.css:676]: Should be fixed now, I think.

Comment (shiftinv) [reply docs/_static/style.css]: updated slightly in https://github.com/DisnakeDev/disnake/pull/963/commits/3107931fe64bc45a829f2c62e1a9c4f730c4166d as discussed

**Pull request merged**